src/limb.rs

```diff
@@ -0,0 +1,271 @@
+use {rand, polyfill, c, core};
+
+// XXX: Not correct for x32 ABIs.
+#[cfg(target_pointer_width = "64")] pub type Limb = u64;
+#[cfg(target_pointer_width = "32")] pub type Limb = u32;
+#[cfg(target_pointer_width = "64")] pub const LIMB_BITS: usize = 64;
+#[cfg(target_pointer_width = "32")] pub const LIMB_BITS: usize = 32;
+
+/// Stores a positive integer range (e.g. that governing a private key) as
+/// little-endian limbs representing the top end of the range (exclusive).
+pub struct Range<'a> {
+    limbs: &'a [Limb],
+}
+
+impl <'a>Range<'a> {
+
+    pub fn from_limbs(limbs: &[Limb]) -> Result<Range, ()> {
+        let r = Range { limbs: limbs };
+        if r.is_valid() {
+            Ok(r)
+        } else {
+            Err(())
+        }
+    }
+
+    fn most_significant_limb(&self) -> Limb {
+        self.limbs[self.limbs.len() - 1]
+    }
+
+    fn leading_zero_bits(&self) -> usize {
+        for i in 0..(LIMB_BITS - 1) {
+            let bit = 1 << (LIMB_BITS - 1 - i);
+            if self.most_significant_limb() & bit > 0 {
+                return i;
+            }
+        }
+
+        unreachable!();
+    }
+
+    fn is_valid(&self) -> bool {
```

Comment:
This should just be inlined into the constructor. Also, all of these checks should be made `debug_assert!`s. This code will only be used internally and only when we know that the range meets all the conditions here.